### PR TITLE
Update input overlay widget updates

### PR DESCRIPTION
**Pull request by mousewax**
Adds the min and max to the input label.
Fields can specify a min, max, or both. 
Values that are not numbers evaluate to 0.

If the number entered by a player or supplied in the JSON/functions is greater than the max, it is adjusted to the max.
If the number entered by a player or supplied in the JSON/functions is less than the min, it is adjusted to the min.
Example JSON:
```
"fields": [
  {
    "type": "number",
    "label": "Which number do you like?",
    "variable": "myNumber",
    "value": 1500,
    "min": 1,
    "max": 100
  },
```

When the field is displayed, it will have a value of 100 and not 1500.

**Bonus fix: If a header is specified, it is inserted as a title.** 
With the new update, headers were ignored in favor of a `title` field. This is to maintain backwards compatibility.

This pull request addresses https://github.com/ArnoldSmith86/virtualtabletop/issues/1008

image

### CODE REVIEW

**Comment by ArnoldSmith86** [comment]
PR-SERVER-BOT: You can play around with it here: http://212.47.248.129:4009/pr-test (or any other room on that server)

**Review by robartsd — DISMISSED** [review]
There is no way to create a number input that does not enforce a range.

**Comment by mousewax** [comment]
> There is no way to create a number input that does not enforce a range.

That is correct. That was true before the update so I kept it.

**Comment by 96LawDawg** [comment]
So we created a breaking change (but not a serious one) with the first INPUT overlay PR.  Changing the "header" parameter to "title" means that the header does not get served up.  Depending on the context, that could be a problem.  Pics below show what Reward recall button prompt used to look like before #976 and after.  It's missing the "Recalling yellow tokens ....." header.

Reward 1

Reward 2

This PR modifies that even further by overwriting what is there with some default text.  It should serve up the original header text.  See what it looks like under this PR:

Reward 3

Recommend we get this fixed and approved soonish (or revert #976) due to the problem of not showing header text that may be important for players.  My fault for not catching this during my review.

**Comment by robartsd** [comment]
> > There is no way to create a number input that does not enforce a range.
> 
> That is correct. That was true before the update so I kept it.

Yes, but limits were not enforced before.

**Comment by 96LawDawg** [comment]
> > There is no way to create a number input that does not enforce a range.
> 
> That is correct. That was true before the update so I kept it.

No, that is not correct.  The button on the right in the test room only allows input between 1 and 10 even though there are no min/max constraints on the button.  That was not the behavior before.

**Review by 96LawDawg — CHANGES_REQUESTED** [review]
Probably this is the min/max problem in lines 406 and 407 of widget.js:
`const inputmin = field.min !== undefined ? field.min : 1;`
`const inputmax = field.max !== undefined ? field.max : 10;`

Maybe those should be null instead of 1 and 10.

**Comment by mousewax** [comment]
I know where the issue is. I did not add those lines. They predate my makeover.

**Comment by mousewax** [comment]
Things have been updated. Test away. If someone could, please dig into the history of that file. I did not add the lines that enforced the min/max. It did indeed work that way before I touched it. Or it should have as that's what it was programmed to do. 

Beyond that, let me know if there are any other updates I should make to this.

**Comment by mousewax** [comment]
> Probably this is the min/max problem in lines 406 and 407 of widget.js: `const inputmin = field.min !== undefined ? field.min : 1;` `const inputmax = field.max !== undefined ? field.max : 10;`
> 
> Maybe those should be null instead of 1 and 10.

I tracked it down. Those lines were committed on 13 July 2021. Regardless of how you think the number input was functioning, it has had this limitation for AT LEAST five months. 

https://github.com/ArnoldSmith86/virtualtabletop/pull/671/files

**Comment by 96LawDawg** [comment]
> 
> I tracked it down. Those lines were committed on 13 July 2021. Regardless of how anyone thinks the number input was functioning, it has had this limitation for AT LEAST five months.
> 
> https://github.com/ArnoldSmith86/virtualtabletop/pull/671/files

Good find!  However, despite those lines being present, for whatever reason, they weren't being enforced.  Go to a PR between July and this week that is still pending and put in an INPUT.  The values are not constrained by default.  Whatever you originally did in this PR made those lines operative.  Enough git blame.

**Comment by 96LawDawg** [comment]
Header parameter now works fine.

Number does not.  If you do not specify min and mix, you are still limited to 1-10.  If you specify min and max, it overrides that default and works fine.  So something is still going on with the default min and max.

**Comment by 96LawDawg** [comment]
So there is another problem with this PR.  It does something really new to the JSON Editor when you try to duplicate a widget.  The ranges listed on the screenshot are new.  Never seen that before.  It's not part of this commit, so something from somewhere else is being brought in.

**Comment by robartsd** [comment]
I recommend removing the min max check code from this PR and just fix the header issue here. We can deal with min/max of numeric input separately.

**Comment by mousewax** [comment]
Looks like the JSON editor uses the same renderer as 

> I recommend removing the min max check code from this PR and just fix the header issue here. We can deal with min/max of numeric input separately.

Beyond the JSON editor, is there any other reason why? Because I just fixed that.

**Comment by 96LawDawg** [comment]
> > I recommend removing the min max check code from this PR and just fix the header issue here. We can deal with min/max of numeric input separately.
> 
> Beyond the JSON editor, is there any other reason why? Because I just fixed that.
> 

Because min/max still doesn't work?

**Comment by mousewax** [comment]
What doesn't work about it? the PR test server seems to be functioning appropriately. What am I missing?

**Comment by mousewax** [comment]
Completed checking on both ends. You can now have a min, max, both, or neither. 

When neither is selected, leading zeros are not stripped, but that shouldn't be a problem for maths done to it.

**Review by 96LawDawg — APPROVED** [review]
Approved through test.  Everything works as it should.

**Comment by rogerl50** [comment]
Screen Shot 2021-12-11 at 8 14 14 PM
The two warnings should not be warnings...

**Review by 96LawDawg — CHANGES_REQUESTED** [review]
Well, it was working until the error-checking popped up.  I think that is overkill.  We aren't letting them design nuclear weapons in the JSON (are we?).  Let the min and max statements that are already there and working be enough.  Recommend changing it back to what it was without the error checking.

**Comment by mousewax** [comment]
> Screen Shot 2021-12-11 at 8 14 14 PM
> 
> The two warnings should not be warnings...

I just put in a fix. By default number inputs are for integers only so decimals are not valid. I added `step="any"` so that decimals no longer trigger a validation problem.

**Comment by mousewax** [comment]
This error is caused when a clickRoutine gets canceled half-way through. The canceled dialog still goes to the next part of the routine but with no variables and thigs get wonky. 

When a `routine` pops up a dialog and they click the cancel button, what should happen with the rest of the `routine`? If it continues, there's nothing passed to it. 

Also, I rolled back to just before #976 was integrated and the error is still there. #976 did not cause this.

**Review by 96LawDawg — APPROVED** [review]
Error checking works fine now.  Recommend approval.

File: client/js/domhelpers.js [hunk 1]
```diff
@@ -35,7 +35,7 @@ export function formField(field, dom, id) {
   const label = document.createElement('label');
   dom.appendChild(label);
   label.htmlFor = id;
-  label.textContent = field.label || '';
+  label.innerHTML = field.label || '';
```
Comment (ArnoldSmith86) [inline client/js/domhelpers.js:38]: Please don't set `innerHTML` to user-provided data.

**Review by ArnoldSmith86 — DISMISSED** [review]
See comment.

File: client/js/domhelpers.js [hunk 2]
```diff
@@ -72,11 +72,31 @@ export function formField(field, dom, id) {
...
+    input.step = "any";
+    input.value = field.value !== undefined ? field.value : 0;
+    if(input.value > field.max){
+      input.value = field.max;
+    }
+    if(input.value < field.min){
+      input.value = field.min;
+    }
+    if(field.min && field.max){
+      label.innerHTML += " <span class='numberInputRange'>("+ field.min +" - "+ field.max +")</span>";
```
Comment (ArnoldSmith86) [inline client/js/domhelpers.js:86]: Still putting user data into `innerHTML`.

Comment (mousewax) [reply client/js/domhelpers.js]: Since those could only be numbers I didn't see a security issue. But I did change it again. Should be fine now.

**Comment by 96LawDawg** [comment]
I just tested everything again.  Still works as expected.  That is, everything works except the known CALL bug.

File: client/js/widgets/widget.js [hunk 3]
```diff
@@ -1725,10 +1735,13 @@ export class Widget extends StateManaged {
       const maxRandomRotate = o.randomRotation || 0;
       const rotation = Math.floor(Math.random() * maxRandomRotate) - (maxRandomRotate / 2);
       var confirmButtonText, cancelButtonText = "";
-
       $('#buttonInputOverlay .modal').style = o.css || "";
       $('#buttonInputOverlay .modal').style.transform = "rotate("+rotation+"deg)";
-      $('#buttonInputFields').innerHTML = '';
+      if(o.header){
+        $('#buttonInputFields').innerHTML = '<div class="inputtitle"><label>'+o.header+'</label></div>';
```
Comment (ArnoldSmith86) [inline client/js/widgets/widget.js:1741]: Still setting user data to `innerHTML`.

File: client/css/jsonedit.css [hunk 4]
```diff
@@ -336,3 +336,10 @@ body.jsonEdit.jeZoomOut #topSurface {
 #widget_dropOffsetY::before{
   content: '[width]';
 }
+#jeCommands .numberInputRange{
+  display: none;
+}
+
+#jeCommands #jeCommandOptions input[type="number"]{
+  width: 6ch;
+}
```
Comment (robartsd) [inline client/css/jsonedit.css:345]: Our regular coding style includes a blank line at the end of files.

File: client/css/layout.css [hunk 5]
```diff
@@ -1070,3 +1070,16 @@ input.switchbox[type="checkbox"]:checked + .switchbox::before {
...
+input:invalid+span:after {
+  font-family: 'Material Icons';
+  content: 'error';
+  position: absolute;
+  right: 5px;
+  top: 0px;
+  font-size: 1.5rem;
+  color: red;
+  letter-spacing: -1px;
+}
```
Comment (robartsd) [inline client/css/layout.css:1085]: Add newline to end.

**Comment by ArnoldSmith86** [comment]
Sorry, this was causing me pain:

```Javascript
    var maxset = false;
    var minset = false;
    if (field.max !== undefined) {
      maxset = true;
    }
    if (field.min !== undefined) {
      minset = true;
    }
```

**Review by ArnoldSmith86 — APPROVED** [review]
Changes look good to me.

**Comment by mousewax** [comment]
All good. I am a simple coder with simple skills.

**Pull request merged**